secure/key-manager/src/key_manager.rs

```diff
@@ -0,0 +1,161 @@
+pub const CONSENSUS_KEY: &str = "consensus_key";
+pub const VALIDATOR_KEY: &str = "validator_key";
+
+impl<'a> KeyManager<'a> {
+    pub fn new(storage: &'a mut dyn Storage) -> Self {
+        Self { storage }
+    }
+
+    /// Initializes the consensus and validator keys for each validator node in secure storage. If
+    /// any of these keys already exist, this method returns an error.
+    pub fn initialize_keys(&mut self) -> Result<(), Error> {
+        // TODO(joshlind): generate the correct policies for these keys (i.e., policies that respect
+        // the LSR and KeyManagement access rules)
+        let consensus_key_policy = Policy::public();
+        let validator_key_policy = Policy::public();
+
+        self.initialize_key(VALIDATOR_KEY, &validator_key_policy)?;
+        self.initialize_key(CONSENSUS_KEY, &consensus_key_policy)
+    }
+
+    /// Initializes a cryptographic key in storage using a given 'key_name' and 'policy'. This
+    /// method verifies the key doesn't already exist and returns an error if the key has already
+    /// been initialized.
+    fn initialize_key(&mut self, key_name: &str, policy: &Policy) -> Result<(), Error> {
+        self.verify_key_not_created(key_name)?;
+        self.storage
+            .generate_new_ed25519_key_pair(key_name, policy)?;
+        Ok(())
+    }
+
+    /// Verifies that a named key has not already been created and stored in secure storage. If the
+    /// key does exist, an error is returned.
+    fn verify_key_not_created(&mut self, key_name: &str) -> Result<(), Error> {
+        match self.storage.get_public_key_for_name(key_name) {
+            Err(Error::KeyNotSet(_)) => Ok(()),
+            _ => Err(Error::KeyAlreadyExists(key_name.to_string())),
+        }
+    }
+
+    /// Rotates the consensus key by: (i) generating and storing a new consensus key in the
+    /// secure store; and (ii) generating and signing a transaction that updates the consensus key
+    /// for this validator node on the blockchain. If a failure occurs during the process (e.g.,
+    /// the consensus key does not exist, or a new transaction could not be generated), an error is
+    /// returned.
+    pub fn rotate_consensus_key(&mut self) -> Result<(), Error> {
+        // Ensure key has already been initialized before rotating
+        self.verify_key_correctly_initialized(CONSENSUS_KEY)?;
+
+        // Rotate the key in storage
+        let new_public_consensus_key = self.storage.rotate_key_pair(CONSENSUS_KEY)?;
+
+        // Update the on-chain validator config to reflect the new_public_consensus_key
+        let _ = self.update_on_chain_consensus_key(new_public_consensus_key);
+
+        Ok(())
+    }
+
+    /// Verifies that a named key has already been initialized correctly in the secure store.
+    /// Returns an error if the key doesn't exist, or if the value related to the key is not
+    /// correct.
+    fn verify_key_correctly_initialized(&mut self, key_name: &str) -> Result<(), Error> {
```

Comment:
merge this in unless there's real reuse. I'm of the opinion that  jumping around to read a snippet of code this small hurts readability and increases total lines of code.